BOOKKEEPER-950: Ledger placement policy to accomodate different storage capacity of bookies

rithin-shetty opened the pull request:
…ge capacity of bookies

This change introduces Disk weight based ledger placement. Currently free disk space is the only supported
weight for a bookie. This change also introduces a new protocol message between bk client and server
called GET_BOOKIE_INFO. This message is used by the client to retrieve the free disk space info from
all the bookies. The existing placement policies: DefaultPlacementPolicy and RackAwarePlacementPolicy
have been enhanced to make use of the weight while selecting bookies. New test cases have been added to
test RackawarePlacement with weights. A new test class has been added to test the weight based selection
algorithm in a stand alone fashion.

sijie commented:
@rithin-shetty  sorry for late response. I will try to get to this one this week.

btw this pull request has conflicts on bookkeeper-server/pom.xml. do you mind rebasing it?

rithin-shetty commented:
@sijie thanks for pointing that out. I have now rebased and updated the branch. The merge conflict should be gone....

rithin-shetty commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/EnsemblePlacementPolicy.java:
```diff
@@ -135,4 +136,12 @@ public EnsemblePlacementPolicy initialize(ClientConfiguration conf,
+     *          A map that has the bookie to BookieInfo
+     */
+    public void updateBookieInfo(Map<BookieSocketAddress, BookieInfo> bookieInfoMap);
 }
```
Sure, I'll add a default implementation.

rithin-shetty commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java:
```diff
@@ -75,6 +75,11 @@
     protected final static String BOOKIE_QUARANTINE_TIME_SECONDS = "bookieQuarantineTimeSeconds";
 
+    // Bookie info poll interval
+    protected final static String DISK_WEIGHT_BASED_PLACEMENT_ENABLED = "diskWeightBasedPlacementEnabled";
```
What is your use case ? Do you want to retrieve the status from all the bookies just once or on a periodic basis ? I'd suggest when you extend the usage of this feature, make the necessary adjustments. You'll have a clearer idea of what the requirements are.

eolivelli replied:
You are right, when we will need to have this info for other purposes we will introduce new flags.

My idea was just to be able to ask for bookie status without altering how the placement policy works, but actually I have to valid use case.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieInfoReader.java:
```diff
@@ -0,0 +1,263 @@
+    private final AtomicBoolean refreshBookieList = new AtomicBoolean();
+
+    public static class BookieInfo implements WeightedObject {
+        private long freeDiskSpace;
```
final?

rithin-shetty replied:
ok. modified...

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookieInfoReader.java:
```diff
@@ -0,0 +1,263 @@
+
+    public static class BookieInfo implements WeightedObject {
+        private long freeDiskSpace;
+        private long totalDiskSpace;
```
final?

rithin-shetty replied:
ok. made final...

rithin-shetty commented:
@sijie I've addressed your comments. I've uploaded the merged changes. But the Jenkins build has been failing with the following error for some time now. Do you know how to fix this ? It's very annoying and should be fixed.

[ERROR] Failed to execute goal org.apache.rat:apache-rat-plugin:0.7:check (default-cli) on project bookkeeper-server: Too many unapproved licenses: 2 -> [Help 1]

sijie commented:
@rithin-shetty I am fixing that by #103

sijie commented:
LGTM. +1

@eolivelli since you were also reviewing this change, can you take a look at the updated pull request?

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java:
```diff
@@ -1720,6 +1730,74 @@ Options getOptions() {
+            return lOpts;
+        }
+
+        String getReadable(long val) {
```
maybe this could be 'static' ?

rithin-shetty replied:
since this is a nested class, I can't make it static.

sijie commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/GetBookieInfoProcessorV3.java:
```diff
@@ -0,0 +1,71 @@
+package org.apache.bookkeeper.proto;
```
need an apache license header

rithin-shetty replied:
ok, will add.

rithin-shetty commented:
@eolivelli I have added a new test case for testing the timeout of GET_BOOKIE_INFO op. See TestGetBookieInfoTimeout.java

jiazhai commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/client/DefaultEnsemblePlacementPolicy.java:
```diff
@@ -51,18 +66,43 @@
+        try {
             allBookies = new ArrayList<BookieSocketAddress>(knownBookies);
+        } finally {
+            rwLock.readLock().unlock();
```
Do we need to consider the race that  'knownBookies' is changed between this line and line 79?

rithin-shetty replied:
Thats should be ok because if we end up choosing a bookie that just went down, the write will fail and we'll come back and change the ensemble. BTW, this could happen even without my change. So I'm not changing the situation any way.

eolivelli commented:
For me is ok
+1

rithin-shetty commented:
@sijie, I have merged and uploaded the latest diffs.
